Makefile

```diff
@@ -14,26 +14,73 @@
 #  limitations under the License.
 #
 
-PROJECT_ROOT 	:= $(shell git rev-parse --show-toplevel)
+ROOT_DIR 	:= $(shell dirname $(realpath $(firstword $(MAKEFILE_LIST))))
+PROTO_TYPE_SUBDIRS = core serving types storage
+PROTO_SERVICE_SUBDIRS = core serving
 
-test:
-	mvn test
+# General
+
+format: format-python lint-go lint-java
+
+lint: lint-python lint-go lint-java
+
+test: test-python test-java
+
+protos: compile-protos-go compile-protos-python compile-protos-docs
+
+build: protos build-java build-docker build-html
 
-test-integration:
-	$(MAKE) -C testing/integration test-integration TYPE=$(TYPE) ID=$(ID)
+# Java
 
-build-proto:
-	$(MAKE) -C protos gen-go
-	$(MAKE) -C protos gen-python
-	$(MAKE) -C protos gen-docs
+format-java:
+	mvn spotless:apply
 
-build-cli:
-	$(MAKE) build-proto
-	$(MAKE) -C cli build-all
+lint-java:
+	mvn spotless:check
+
+test-java:
+	mvn test
 
 build-java:
 	mvn clean verify
 
+# Python SDK
+
+install-python-ci-dependencies:
+	pip install -r sdk/python/requirements-ci.txt
+
+compile-protos-python: install-python-ci-dependencies
+	@$(foreach dir,$(PROTO_TYPE_SUBDIRS),cd ${ROOT_DIR}/protos; python -m grpc_tools.protoc -I. --python_out=../sdk/python/ --mypy_out=../sdk/python/ feast/$(dir)/*.proto;)
+	@$(foreach dir,$(PROTO_SERVICE_SUBDIRS),cd ${ROOT_DIR}/protos; python -m grpc_tools.protoc -I. --grpc_python_out=../sdk/python/ feast/$(dir)/*.proto;)
+	cd ${ROOT_DIR}/protos; python -m grpc_tools.protoc -I. --python_out=../sdk/python/ --mypy_out=../sdk/python/ tensorflow_metadata/proto/v0/*.proto
+
+test-python:
+	pytest --verbose --color=yes sdk/python/tests
+
+format-python:
+	cd ${ROOT_DIR}/sdk/python; isort -rc feast tests
+	cd ${ROOT_DIR}/sdk/python; black --target-version py37 feast tests
+
+lint-python:
+	# TODO: This mypy test needs to be re-enabled and all failures fixed
+	#cd ${ROOT_DIR}/sdk/python; mypy feast/ tests/
+	cd ${ROOT_DIR}/sdk/python; flake8 feast/ tests/
+	cd ${ROOT_DIR}/sdk/python; black --check feast tests
+	cd ${ROOT_DIR}/sdk/python; isort -rc feast tests --check-only
+
+# Go SDK
+
+compile-protos-go:
+	@$(foreach dir,$(PROTO_TYPE_SUBDIRS), cd ${ROOT_DIR}/protos; protoc -I/usr/local/include -I. --go_out=plugins=grpc,paths=source_relative:../sdk/go/protos/ feast/$(dir)/*.proto;)
+
+format-go:
+	cd ${ROOT_DIR}/sdk/go; gofmt -s -w *.go
+
+lint-go:
+	cd ${ROOT_DIR}/sdk/go; go vet; golint *.go
```

Comment:
If my version of `golint` is still current, it requires a flag `-set_exit_status` in order to return an error status to the shell.

A formatting check should probably be included in `lint-go` as well, it is for the other languages. `gofmt` has no such feature at all to return an error exit status, the idiom seems to be to use `test -z $(gofmt -l ./)` when this is desired.

I'm not sure if we're actually wanting it to fail on CI yet, because of the desire to not fail lint checks one-at-a-time, but we probably want to be consistent about whether all the `lint-*` make targets exit nonzero when there's an issue (I haven't checked behavior of all others yet).

Probably `lint-go` will not fail CI for now even if this target is updated to exit nonzero, because `.prow/scripts/test-golang-sdk.sh` does not `set -e`.